src/libfsm/intersect.c

```diff
@@ -65,3 +72,317 @@ fsm_intersect(struct fsm *a, struct fsm *b)
+		next = p->next;
+		free(p);
+	}
+}
+
+static struct pair *alloc_pair(struct bywalk_arena *A)
+{
+	static const struct pair zero;
+
+	struct pair *item;
+	struct pair_pool *pool;
+
+	if (A->head == NULL) {
+		goto new_pool;
+	}
+
+	if (A->top >= (sizeof A->head->items / sizeof A->head->items[0])) {
+		goto new_pool;
+	}
+
+	item = &A->head->items[A->top++];
+	*item = zero;
+	return item;
+
+new_pool:
+	pool = malloc(sizeof *pool);
+	if (pool == NULL) {
+		return NULL;
+	}
+
+	pool->next = A->head;
+	A->head = pool;
+	A->top = 1;
+
+	item = &pool->items[0];
+	*item = zero;
+	return item;
+}
+
+static struct pair *new_pair(struct bywalk_arena *A, struct fsm_state *a, struct fsm_state *b)
+{
+	struct pair lkup, *p;
+	struct fsm_state *combined;
+
+	lkup.a = a;
+	lkup.b = b;
+
+	assert(A->states);
+
+	p = set_contains(A->states, &lkup);
+	if (p != NULL) {
+		return p;
+	}
+
+	p = alloc_pair(A);
+	if (p == NULL) {
+		return NULL;
+	}
+
+	combined = fsm_addstate(A->new);
+	if (combined == NULL) {
+		return NULL;
+	}
+	combined->equiv = NULL;
+
+	p->a = a;
+	p->b = b;
+	p->comb = combined;
+	if (!set_add(&A->states, p)) {
+		return NULL;
+	}
+
+	if (a->end && b->end) {
+		const struct fsm_options *opt;
+
+		fsm_setend(A->new, combined, 1);
+
+		opt = A->new->opt;
+		if (opt != NULL && opt->carryopaque != NULL) {
+			const struct fsm_state *states[2];
+			states[0] = a;
+			states[1] = b;
+			/* this is slightly cheesed, but it avoids
+			 * constructing a set just to pass these two
+			 * states to the carryopaque function
+			 */
+			opt->carryopaque(states, 2, A->new, combined);
+		}
+	}
+
+	return p;
+}
+
+static int
+intersection_walk_edges(struct bywalk_arena *A, struct fsm *a, struct fsm *b, struct pair *start);
+
+static void
+mark_equiv_null(struct fsm *fsm);
+
+struct fsm *
+fsm_intersect_bywalk(struct fsm *a, struct fsm *b)
+{
+	/* intersection via DFS of the two DFAs.
+	 *
+	 * We do this in a few steps:
+	 *
+	 *   1. Initialize things:
+	 *        a) Initialize state map, which maps (q0,q1) pairs to
+	 *           new states.
+	 *        b) Let start0,start1 be the starting states of the two
+	 *           DFAs.  Allocate a new combined state for
+	 *           (start0,start1) and add the combined state to the
+	 *           queue.
+	 *
+	 *        c) We use the internal state equiv field as a
+	 *           'visited' marker.  If st->equiv == NULL, the
+	 *           state has not yet been visited.
+	 */
+
+	static const struct bywalk_arena zero;
+	struct bywalk_arena A = zero;
+
+	struct fsm *new = NULL;
+	struct fsm_state *sa, *sb;
+	struct pair *pair0;
+
+	assert(a != NULL);
+	assert(b != NULL);
+
+	A.new = fsm_new(a->opt);
+	if (A.new == NULL) {
+		goto error;
+	}
+
+	A.states = set_create(cmp_pair);
+	if (A.states == NULL) {
+		goto error;
+	}
+
+	sa = fsm_getstart(a);
+	sb = fsm_getstart(b);
+
+	if (sa == NULL || sb == NULL) {
+		/* intersection will be empty.  XXX - should this be an error? */
```

Comment:
I'd argue no - because we're considering FSMs defined by the language they accept. The language here is the empty set. So the intersection will also be the empty set.

That's output as an FSM with a single start state, and it's not marked as an end state. `fsm_minimise()`, `fsm_reverse()` and friends now produce this. (Perhaps we ought to centralise something which produces it)